src/solvers/smt2/smt2_parser.cpp

```diff
@@ -699,44 +728,80 @@ exprt smt2_parsert::function_application()
           if(next_token() != smt2_tokenizert::CLOSE)
             throw error("expected ')' after to_fp");
 
-          auto rounding_mode = expression();
+          // width_f *includes* the hidden bit
+          const ieee_float_spect spec(width_f - 1, width_e);
 
-          if(next_token() != smt2_tokenizert::NUMERAL)
-            throw error("expected number after to_fp");
+          auto rounding_mode = expression();
 
-          auto real_number = smt2_tokenizer.get_buffer();
+          auto source_op = expression();
 
           if(next_token() != smt2_tokenizert::CLOSE)
             throw error("expected ')' at the end of to_fp");
 
-          mp_integer significand, exponent;
-
-          auto dot_pos = real_number.find('.');
-          if(dot_pos == std::string::npos)
+          // There are several options for the source operand:
+          // 1) real or integer
+          // 2) bit-vector, which is interpreted as signed 2's complement
+          // 3) another floating-point format
+          if(
+            source_op.type().id() == ID_real ||
+            source_op.type().id() == ID_integer)
           {
-            exponent = 0;
-            significand = string2integer(real_number);
+            // For now, we can only do this when
+            // the source operand is a constant.
+            if(source_op.id() == ID_constant)
+            {
+              mp_integer significand, exponent;
+
+              const auto &real_number =
+                id2string(to_constant_expr(source_op).get_value());
+              auto dot_pos = real_number.find('.');
+              if(dot_pos == std::string::npos)
+              {
+                exponent = 0;
+                significand = string2integer(real_number);
+              }
+              else
+              {
+                // remove the '.', if any
+                std::string significand_str;
+                significand_str.reserve(real_number.size());
+                for(auto ch : real_number)
+                  if(ch != '.')
+                    significand_str += ch;
+
+                exponent =
+                  mp_integer(dot_pos) - mp_integer(real_number.size()) + 1;
+                significand = string2integer(significand_str);
+              }
+
+              ieee_floatt a(spec);
+              a.rounding_mode = static_cast<ieee_floatt::rounding_modet>(
+                numeric_cast_v<int>(to_constant_expr(rounding_mode)));
```

Comment:
We really ought to have a constructor of `ieee_floatt` that takes a rounding mode...